src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -292,6 +297,48 @@ private static RouteHandlerFilterDelegate CreateFilterPipeline(MethodInfo method
         return filteredInvocation;
     }
 
+    private static Expression MapHandlerReturnTypeToValueTask(Expression methodCall, Type returnType)
+    {
+        if (returnType == typeof(void) || returnType == typeof(Task) || returnType == typeof(ValueTask))
+        {
+            return Expression.Block(methodCall, Expression.Constant(new ValueTask<object?>(EmptyHttpResult.Instance)));
+        }
+        else if (returnType.IsGenericType &&
+                     returnType.GetGenericTypeDefinition() == typeof(ValueTask<>))
+        {
+            var typeArg = returnType.GetGenericArguments()[0];
+            return Expression.Call(ValueTaskOfTToValueTaskOfObjectMethod.MakeGenericMethod(typeArg), methodCall);
```

Comment:
This works, right? This would avoid an async state machine allocation per request at runtime I think for route handlers that are returning ValueTasks of reference types.

```suggestion
            if (typeArg.IsValueType)
            {
                return Expression.Call(ValueTaskOfTToValueTaskOfObjectMethod.MakeGenericMethod(typeArg), methodCall);
            }
            
            return methodCall;
```